File: DotNet/Messaging/MessageFetchers/HydraByTopicMessageFetcher.cs
using Shastra.Hydra.Messaging.MessageIds;
using Shastra.Hydra.Messaging.Storage;

namespace Shastra.Hydra.Messaging.MessageFetchers
{
    public class HydraByTopicMessageFetcher : MessageFetcherBase<HydraMessage>
    {
        private readonly string _topic;
        private readonly IKeyOptions _endKey;

        public HydraByTopicMessageFetcher(string topic)
        {
            _topic = topic;
            _endKey = new KeyOptions(_topic, KeyOptions.MaxValue);
        }

        protected override string ViewName { get { return "broadcastMessages"; } }

        // The broadcastMessages view is indexed on [topic, id]
        protected override IKeyOptions MessageKey(IMessageId id) { return new KeyOptions(_topic, id.ToDocId()); }
        protected override IKeyOptions EndKey() { return _endKey; }

    }
}

File: Java/hydra/src/main/java/uk/co/shastra/hydra/messaging/messagefetchers/HydraByTopicMessageFetcher.java
package uk.co.shastra.hydra.messaging.messagefetchers;

import org.ektorp.ComplexKey;

import uk.co.shastra.hydra.messaging.HydraMessage;
import uk.co.shastra.hydra.messaging.messageids.MessageId;

public class HydraByTopicMessageFetcher extends MessageFetcherBase<HydraMessage> {

	private String topic;
	private ComplexKey endKey;
    
	public HydraByTopicMessageFetcher(String topic) {
		super(HydraMessage.class);
		this.topic = topic;
		// TODO - check that "{}" is a suitable max value
		endKey = ComplexKey.of(topic, ComplexKey.emptyObject());
	}

	@Override
	protected String getViewName() { return "broadcastMessages"; }

    // The broadcastMessages view is indexed on [topic, id]
	@Override
	protected ComplexKey getMessageKey(MessageId id) { return ComplexKey.of(topic, id.toDocId()); }

	@Override
	protected ComplexKey getEndKey() { return endKey; }

}
